modules/photons_control/clean.py

```diff
@@ -0,0 +1,188 @@
+
+        if "indication" not in options:
+            options["indication"] = False
+
+        if "duration" not in options:
+            options["duration"] = 7200
+
+        await self.target.send(CleanConfig(**options), self.reference, **kwargs)
+
+
+@task
+class get_clean_default_config(task.Task):
+    """
+    Get the default HEV cleaning cycle configuration for the target bulb
+    """
+
+    target = task.requires_target()
+    reference = task.provides_reference(special=True)
+
+    async def execute_task(self, **kwargs):
+        async def gen(reference, sender, **kwargs):
+            ps = sender.make_plans("capability")
+            async for serial, _, info in sender.gatherer.gather(ps, reference, **kwargs):
+                if info["cap"].has_hev:
+                    yield LightMessages.GetHevCycleConfiguration(target=serial)
+
+        async for pkt in self.target.send(FromGenerator(gen), self.reference):
+            print(pkt.serial)
+            indication = "enabled" if pkt.indication else "disabled"
+            print(f"   Cycle time: {pkt.duration_s}")
+            print(f"\tIndication: {indication}")
+            print("")
+
+
+def CleanCycle(enable=None, duration=0, **kwargs):
+    """
+    Returns a valid message that will either start or stop the cleaning (HEV)
+    cycle of devices used against it.
+
+    When a cycle is started, if the duration=0 or is not provided, the default
+    duration will be used.
+
+    For example:
+
+    .. code-block:: python
+
+        await target.send(CleanCycle(), ["d073d5000001", "d073d5000001"])
+    """
+
+    async def gen(reference, sender, **kwargs):
+        if enable is not None:
+            has_hev = DeviceMessages.GetVersion()
+            async for pkt in sender(has_hev, reference, **kwargs):
+                if pkt | DeviceMessages.StateVersion:
+                    if pkt.product == 90 or pkt.product == 99:
+                        yield LightMessages.SetHevCycle(
+                            enable=enable,
+                            res_required=False,
+                            duration_s=duration,
+                            target=pkt.serial,
+                        )
+
+    return FromGenerator(gen)
+
+
+@task
+class start_clean_cycle(task.Task):
+    """
+    Start a cleaning cycle on an HEV-capable bulb
+
+    ``target:start_clean_cycle match:cap=hev -- '{"duration": 7200}'``
+
+    It  takes a ``duration`` field that is the seconds of the duration. If not
+    provided, or set to 0, the default duration will be used.
+    """
+
+    target = task.requires_target()
+    reference = task.provides_reference(special=True)
+
+    async def execute_task(self, **kwargs):
+        options = self.photons_app.extra_as_json
+        options["enable"] = True
+
+        if "duration" not in options:
+            options["duration"] = 0
+
+        await self.target.send(CleanCycle(**options), self.reference)
+
+
+@task
+class stop_clean_cycle(task.Task):
```

Comment:
The class system means you could factor out most of these tasks.

So something like

```
class change_clean_cycle(Task):
    target = ...
    reference = ...

    defaults = dictobj.Field(sb.overridden({}))

    async def execute_task(self):
        options = MergedOptions.using(self.defaults, self.photons_app.extra_as_json).as_dict()
        await self.target.send(CleanCycle(**options), self.reference)

class stop_clean_cycle(change_clean_cycle):
    defaults = dictobj.Field(sb.overridden({"enable": False, "duration": 0}))

class start_clean_cycle(change_clean_cycle):
    defaults = dictobj.Field(sb.overridden({"enable": True, "duration": 7200}))
```

And move the normalisation into the CleanCycle message

Reply:
Can you point me towards any existing tasks that use this pattern (or something resembling it)?

Reply:
think of it like any normal class

```
class Thing:
  a = 1
  def execute(self):
      print(self.a)

class Other(Thing):
    a = 2

class Stuff(Thing):
    a = 3

Thing().execute() # prints 1
Other().execute() # prints 2
Stuff().execute() # prints 3
```

Reply:
The Task is an instance of [dictobj.Spec](https://delfick-project.readthedocs.io/en/latest/api/norms/api/dictobj.html#module-delfick_project.norms.field_spec)

Reply:
D'oh, of course.